### DISCUSSION TITLE
ARMv7 compatibility and weird issues

### DISCUSSION THREAD

**Discussion by chiveicrook2** [tcunit/TcUnit#147] [Q&A] [answered]
Hello,
I've been trying to use TcUnit in my CX8180 project. Unfortunately, I've run into some weird issues and started to wonder if TcUnit is compatible with ARMv7 targets at all. I'm a twincat newbie so I'm not really sure whether the problem is in my libraries, settings or elsewhere and I'm at a loss as to why referencing a library which uses TcUnit causes all kinds of issues - unit testing code shouldn't even execute in such case, right? 

At first, I developed and tested 2 libraries on local x64 twincat runtime and everything seemed ok. The moment I reference my libraries in CX8180 project (even if nothing uses them!) I lose ability to set breakpoints, value forcing is unreliable and PLC seems to randomly skip instructions. 

To troubleshoot the issue I've tried the SequentialTestSuitesAndTests example project which when compiled and activated on CX8180 causes the same issue - even though the tests execute properly it is impossible to set a breakpoint anywhere and flow control doesn't work.

What am I doing wrong?

**Comment by sagatowski** [comment]
Hi!

Which version of TwinCAT are you using?
Which version of TcUnit? Make sure you run a formal release.

That's right, if you don't execute TcUnit (by `TcUnit.RUN()`) the unit test code won't be executed).

Can you try with the simple example project?

Edit: I really can't see how TcUnit (which is just a library) can affect your ability to set the breakpoints. If you remove TcUnit from your libraries, does it work fine then?

**Reply by chiveicrook2** [reply]
Currently I'm on TwinCAT 4024.10 and TcUnit straight from github master, previously I've used 4024.7 and TcUnit 1.1. 
I'll try with a formal 1.1 and 4024.10 and get back to you.

**Reply by chiveicrook2** [reply]
Finished some tests with Twincat 4024.10 + TcUnit 1.1 installed from source (formal zip release):

**1. SimpleTcUnitExampleProject, local TwinCAT RT (x64):**
Result: 2/2 tests succeed. Everything works as expected. 

**2. SimpleTcUnitExampleProject, CX8180 TwinCAT CE7 (ARMv7):**
a) Build -> clean solution
b) Build -> configuration manager -> set active solution platform to TwinCAT CE7 (ARMv7)
c) Set target to my CX
d) Read realtime settings from target
e) Build solution
f) Activate configuration and restart in run mode
Result: 2/2 tests succeed. After login and any attempt to set a breakpoint (anywhere, including right on "TcUnit.RUN()") I get "TwinCAT PLC Control Breakpoint could not be set on target." Error.

**3. New barebones project, CX8180 target, 4024.10:**
```
PROGRAM MAIN
VAR
	i : UINT := 0;
END_VAR
i := i+1;
```
Without referencing TcUnit: Works fine
Referencing any user-created library without TcUnit: Works Fine
Just referencing TcUnit or referencing any library which references TcUnit: 
On first try I got something new: Trying to set a breakpoint results in "Exception 0xc000001d Illegal instruction"
Rebuild and second try: "TwinCAT PLC Control Breakpoint could not be set on target."
Remove library reference and rebuild: Works fine

At first I thought it was some issue with library interaction but now I see that just referencing TcUnit causes problems. I don't see how that's possible but here it is...

**Reply by sagatowski** [reply]
That is indeed weird. I have honestly not used a ARM-based Beckhoff PLC in a while, but there are some users using these type of PLCs with TcUnit, though honestly I'm not sure anyone uses breakpoints.

I'll order a CX8180 and do some testing.

**Reply by chiveicrook2** [reply]
Oh well, I guess I will also do some experiments in the meantime. Inability to set breakpoints is the easiest to reproduce common denominator; for example, in my main project some of the code was always skipped as well. 
Anyway, TcUnit is a dream come true, thank you for your efforts to bring modern development practices to automation world :-)

**Reply by sagatowski** [reply]
Meanwhile while you wait, you can contact your local Beckhoff support and see whether they have any idea?

Happy that you find TcUnit useful!

**Reply by chiveicrook2** [reply]
Well, all I got from local Beckhoff support was a oneliner: 

> ARM and Intel architectures are fundamentally different so unless the library is designed for ARM problems can arise.

I guess I could try contacting global support but most likely they'll just forward me to local support. 

In the meantime I tried different clean Twincat installs in VMs to rule out any external influences and no luck. Same with different versions of clean images for CX.

Search continues :-)

**Reply by sagatowski** [reply]
Hi @chiveicrook2 ! Yes I've received a couple of those one-liners as well. Not much information to go on.

I have a CX8190 on its way to me, though it's probably ~2 weeks away still.
Thanks for keeping us updated!

**Reply by sagatowski** [reply]
@chiveicrook2 I just realized one thing. Can you try to build the library yourself for ArmV7 using the sourcecode? Just clone the latest version and see if it makes any difference.

**Reply by chiveicrook2** [reply]
> @chiveicrook2 I just realized one thing. Can you try to build the library yourself for ArmV7 using the sourcecode? Just clone the latest version and see if it makes any difference.

Actually that's how I started: Just cloned the repo and built it for ARM. Builds just fine, still doesnt work. Even tried to build for ARM and save as compiled library, no difference. I'll try to backtrack to previous versions, maybe one of them will work.

**Reply by chiveicrook2** [reply]
@sagatowski I think I've pinpointed the problem to commit c067847. Commit 5db7b40 is the last one which I can use without any immediate problems. Didn't have time for more thorough tests but with 5db7b40 I can set breakpoints and with c067847 I can't. I tried to apply changes from c067847 to 5db7b40 and ended up with a functional, yet seemingly identical to c067847 project. It's a hefty commit so either I missed something or there is something git doesn't track properly between these two.

**Reply by sagatowski** [reply]
@chiveicrook2 Fantastic detective work! Highly appreciated. I think this should be a "real" issue, so I've created a new issue out from this #148 and we can continue the discussion there.

**Reply by sagatowski** [reply]
@chiveicrook2 I just want to make sure.... it is still possible to run TcUnit with the CX8180, right? You still get all results and everything. It's just that you can't debug your code for some reason?

**Reply by chiveicrook2** [reply]
@sagatowski  That's right, tests execute properly and display expected results despite aforementioned issues (at least in my test suites which aren't particularly complex)

**Comment by sagatowski** [comment, marked as answer]
Answer for this discussion is provided in [the issue that now is closed](https://github.com/tcunit/TcUnit/issues/148).
